### PR TITLE
[WINLOGON] Implement system sounds.

### PR DESCRIPTION
**Pull request by TAN-Gaming**
[WINLOGON] Implement system sounds.

JIRA issue: [CORE-13951](https://jira.reactos.org/browse/CORE-13951)

Working sounds
- Logon sound (improved)
- Logoff sound
- MessageBox and MessageBeep sounds

### CODE REVIEW

File: base/system/winlogon/sas.c [hunk 1]
```diff
@@ -279,153 +279,305 @@ PlaySoundRoutine(
     return Ret;
 }
 
-DWORD
-WINAPI
-PlayLogonSoundThread(
-    IN LPVOID lpParameter)
+DWORD WINAPI PlayLogonSoundThread(LPVOID lpParameter)
```
Comment (HBelusca) [inline base/system/winlogon/sas.c:282]: Please revert.

File: base/system/winlogon/sas.c [hunk 2]
```diff
@@ -279,153 +279,305 @@ PlaySoundRoutine(
...
+
+    hThread = CreateThread(NULL, 0, (LPTHREAD_START_ROUTINE)PlayLogonSoundThread, (LPVOID)Session, 0, NULL);
+
+    if (hThread)
+    {
+        CloseHandle(hThread);
+    }
+}
+
+static void PlayLogoffSound(PWLSESSION Session)
```
Comment (HBelusca) [inline base/system/winlogon/sas.c:571]: There seems to be quite redundancy between these functions, whose behaviour just differ by only **one** thing (the thread function)!

File: base/system/winlogon/sas.c [hunk 3]
```diff
@@ -1344,38 +1497,142 @@ UnregisterHotKeys(
...
-        break;
-    case MB_ICONQUESTION:
-        EventName = L"SystemQuestion";
-        break;
-    default:
-        WARN("Unhandled type %d\n", uType);
-        EventName = L"SystemDefault";
-    }
-
-    return PlaySoundRoutine(EventName, FALSE, SND_ALIAS | SND_NOWAIT | SND_NOSTOP | SND_ASYNC);
```
Comment (HBelusca) [inline base/system/winlogon/sas.c:1378]: What's the rationale behind why it's not possible to directly use the PlaySoundRoutine as is?

File: base/system/winlogon/sas.c [hunk 4]
```diff
@@ -279,153 +279,305 @@ PlaySoundRoutine(
...
+    HeapFree(GetProcessHeap(), 0, lpSndPath);
+
     return 0;
 }
 
-static
-VOID
-PlayLogonSound(
-    IN OUT PWLSESSION Session)
+static void PlayLogonSound(PWLSESSION Session)
```
Comment (HBelusca) [inline base/system/winlogon/sas.c:559]: Please revert.

File: base/system/winlogon/sas.c [hunk 5]
```diff
@@ -279,153 +279,305 @@ PlaySoundRoutine(
...
+    /* Play sound */
+    PlaySoundRoutine(lpSndPath, TRUE, SND_FILENAME | SND_NODEFAULT);
+
+    HeapFree(GetProcessHeap(), 0, lpRegVal);
+    HeapFree(GetProcessHeap(), 0, lpSndPath);
+
+    return 0;
+}
+
+DWORD WINAPI PlayLogoffSoundThread(LPVOID lpParameter)
```
Comment (HBelusca) [inline base/system/winlogon/sas.c:449]: Both logon and logoff functions seem to be quite similar; isn't it possible to merge both together, and have the actual different behaviour be controlled by a parameter?

File: base/system/winlogon/sas.c (hunk 3 above)
Comment (TAN-Gaming) [reply base/system/winlogon/sas.c:1378]: PlaySound function with SND_ASYNC flag calls CreateThread function that doesn't work with user impersonation.

File: base/system/winlogon/sas.c (hunk 1 above)
Comment (TAN-Gaming) [reply base/system/winlogon/sas.c]: Yep

File: base/system/winlogon/sas.c (hunk 4 above)
Comment (TAN-Gaming) [reply base/system/winlogon/sas.c]: Yep

File: base/system/winlogon/sas.c (hunk 2 above)
Comment (TAN-Gaming) [reply base/system/winlogon/sas.c]: Updated.

File: base/system/winlogon/sas.c (hunk 5 above)
Comment (TAN-Gaming) [reply base/system/winlogon/sas.c]: Updated.

**Review by oleg-dubinskiy — COMMENTED** [review]
As you might notice, `PlaySoundRoutine` that is called from these functions, delay-loads `PlaySoundW` from winmm.dll, which actually does all work for playing the system sound events (like done in Windows).
Please note that some of our current winmm implementations are not working correctly (in particular, `get_mmioFromProfile`, which is responsible for `SND_ALIAS` flag, receives an empty string from registry values). Although you didn't call that `SND_ALIAS` flag in your code, nevertheless it's important anyway I think. I'm not sure is it caused by winlogon issues (those might be fixed by your PR), but winmm bugs still might prevent to playing the sound events properly. Btw, I made some attempts to fix that in the past, but for now without luck. :frowning_face: 
See dll/win32/winmm/playsound.c file for the reference: https://git.reactos.org/?p=reactos.git;a=blob;f=dll/win32/winmm/playsound.c;h=a6fdbd7ed195285d9d799936000321cab7aa13f3;hb=HEAD. And feel free to test your changes locally with enabled sound events (set any wav sound files as sound events manually in Control Panel). :wink:

**Comment by TAN-Gaming** [comment]
@oleg-dubinskiy I found that  `get_mmioFromProfile` doesn't work with winlogon because it reads registry values from HKEY_CURRENT_USER (predefined key) so in this case, user impersonation is needed to call `RegOpenCurrentUser` and access user's HKCU

**Comment by TAN-Gaming** [comment]
I've fixed my fork and opened a new PR.

**Pull request closed without merging**